src/kernel/pagecache.c

```diff
@@ -616,49 +614,86 @@ closure_function(1, 3, void, pagecache_write_sg,
+        pagecache_page pp = page_lookup_nodelocked(pn, pi);
+        if (pp == INVALID_ADDRESS) {
+            pp = allocate_page_nodelocked(pn, pi);
+            if (pp == INVALID_ADDRESS) {
+                pagecache_unlock_node(pn);
+                apply(completion, timm("result", "failed to allocate pagecache_page"));
+                return;
+            }
+
+            /* When writing a new page at the end of a node whose length is not block-aligned, zero
+               the remaining portion of the last block. The filesystem will depend on this to properly
+               implement file holes. */
+            range i = range_intersection(byte_range_from_page(pc, pp), q);
+            u64 tail_offset = i.end & MASK(pv->block_order);
+            if (tail_offset) {
+                u64 page_offset = i.end & MASK(pc->page_order);
+                u64 len = U64_FROM_BIT(pv->block_order) - tail_offset;
+                pagecache_debug("   zero unaligned end, i %R, page offset 0x%lx, len 0x%lx\n",
+                                i, page_offset, len);
+                assert(i.end == pn->length);
+                zero(pp->kvirt + page_offset, len);
+            }
+        }
         pagecache_lock_state(pc);
+        if (page_state(pp) == PAGECACHE_PAGESTATE_FREE)
+            realloc_pagelocked(pc, pp);
+        refcount_reserve(&pp->refcount);
         if (page_state(pp) == PAGECACHE_PAGESTATE_READING)
             enqueue_page_completion_statelocked(pc, pp, apply_merge(m));
         pagecache_unlock_state(pc);
-        pp = (pagecache_page)rbnode_get_next((rbnode)pp);
     }
     pagecache_unlock_node(pn);
     apply(sh, STATUS_OK);
 }
 
 /* evict pages from new and active lists, then rebalance */
-static u64 evict_pages_locked(pagecache pc, u64 pages)
+static u64 evict_pages_locked(pagecache pc, u64 pages, vector evictlist)
 {
-    u64 evicted = evict_from_list_locked(pc, &pc->new, pages);
+    u64 evicted = evict_from_list_locked(pc, &pc->new, evictlist, pages);
     if (evicted < pages) {
         /* To fill the requested pages evictions, we are more
            aggressive here, evicting even in-use pages (rc > 1) in the
            active list. */
-        evicted += evict_from_list_locked(pc, &pc->active, pages - evicted);
+        evicted += evict_from_list_locked(pc, &pc->active, evictlist, pages - evicted);
     }
-    balance_page_lists_locked(pc);
     return evicted;
 }
 
+#define DRAIN_ITER_MAX   128
+
 u64 pagecache_drain(u64 drain_bytes)
 {
+    pagecache_page pp;
     pagecache pc = global_pagecache;
     u64 pages = pad(drain_bytes, cache_pagesize(pc)) >> pc->page_order;
+    vector v;
+    u64 evicted = 0;
+    u64 lastevicted = 0;
 
-    /* We could avoid taking both locks here if we keep drained page
-       objects around (which incidentally could be useful to keep
-       refault data). */
+    if ((v = allocate_vector(pc->h, DRAIN_ITER_MAX)) == INVALID_ADDRESS)
+        return 0;
+    while (evicted < pages) {
```

Comment:
Pages is being decremented while evicted is accumulating, so this should either change to pages > 0, or perhaps better yet, remove the pages decrement below and use MIN(pages - evicted, DRAIN_ITER_MAX). Also, keeping lastevicted isn't really necessary - could just check the return value of evict_pages_locked, break if 0, else add to evicted.

Reply:
Nice catch! I corrected and cleaned it up.